wallet: Move salvagewallet into wallettool

achow101 opened the pull request:
Removes the `-salvagewallet` startup option and adds a `salvage` command to the `bitcoin-wallet` tool. As such, `-salvagewallet` is removed. Additionally, the automatic salvage that is done if the wallet file fails to load is removed.

Lastly the salvage code entirely is moved out entirely into `bitcoin-wallet` from `walletdb.{cpp/h}` and `db.{cpp/h}`.

achow101 committed:
wallettool: Add a salvage command

achow101 committed:
wallet: remove -salvagewallet

achow101 committed:
walletdb: don't automatically salvage when corruption is detected

MarcoFalke commented on test/functional/wallet_basic.py:
```diff
@@ -404,8 +404,6 @@ def run_test(self):
             '-zapwallettxes=1',
             '-zapwallettxes=2',
-            # disabled until issue is fixed: https://github.com/bitcoin/bitcoin/issues/7463
-            # '-salvagewallet',
```
Would be nice to keep this test or at least document that salvagewallet will throw away your keys: #7379

MarcoFalke replied:
> @MarcoFalke could you describe the test or comment you would add? It sounds there is a good suggestion here, but I don't understand it's asking for

The new functionality has zero tests. I was hoping we could call `wallet-tool salvage` instead of `bitcoind -salvagewallet` here. Just because the test is commented out, doesn't mean it should be removed.

If the test still fails occasionally with the wallet-tool, then it can be commented out again.

achow101 replied:
The behavior of salvage shouldn't be any different, so it will probably still fail as the previous `-salvagewallet` did.

achow101 replied:
I've added a placeholder test with a todo that mentions the issue in `tool_wallet.py`.

MarcoFalke commented:
Concept ACK

laanwj commented:
Concept ACK

jonasschnelli commented:
Concept ACK, ... finally.
Follows the concept of #8745. There was already an attempt to add salvage to the wallet tool in #15307

ryanofsky reviewed (commented):
Started reviewing 67e30b2b42b4d710739e6fe303cbca69be68032c but finding it difficult to compare old and new behavior with all the code moving and changing at the same time. Also wondering if the ReadKeyValue code duplication can be avoided, because it seems likely to doom this feature to obsolescence if updates to that function need to be mirrored here. It makes me wonder what the plan for salvage wallet is. Is the plan to wall it off, deprecate it, and remove it? It seems like -salvagewallet was implemented in a really convoluted way, but if we did want to implement in a sustainable way, we would still implement it as a load option and not as a standalone copy of loading code. (It could still be exposed through wallet tool either way.)

But overall this seems promising, and great to get rid of some of the convoluted callback & enum status code. I will spend more time with this and maybe try to come up with a MOVEONLY version that could be easier to review.

achow101 committed:
walletdb: remove fAggressive from Salvage

The only call to Salvage set fAggressive = true so remove that parameter
and always use DB_AGGRESSIVE

achow101 committed:
Move BerkeleyEnvironment::Salvage into BerkeleyBatch::Recover

achow101 committed:
Expose a version of ReadKeyValue and use it in RecoverKeysOnlyFilter

We need this exposed for BerkeleyBatch::Recover to be moved out.

achow101 committed:
Make BerkeleyBatch::Recover and WalletBatch::RecoverKeysOnlyFilter standalone

Instead of having these be class static functions, just make them be
standalone. Also removes WalletBatch::Recover which just passed through
to BerkeleyBatch::Recover.

achow101 committed:
Move RecoverDatabaseFile and RecoverKeysOnlyFilter into salvage.{cpp/h}

achow101 committed:
Move RecoverKeysOnlyFilter into RecoverDataBaseFile

achow101 commented on src/wallet/wallettool.cpp:
```diff
@@ -103,11 +104,205 @@ static void WalletShowInfo(CWallet* wallet_instance)
+/* End of key/value data */
+static const char *DATA_END = "DATA=END";
+
 static bool SalvageWallet(fs::path path)
```
I've added a salvage.{cpp/h} and put the salvage stuff in there.

achow101 commented:
I've changed the final refactor commit into several smaller moves and changes. The final approach for salvage is also slightly different.

Instead of having all of the salvage logic `wallettool.cpp`, I've moved instead put it in a `RecoverDatabaseFile` function in new `salvage.{cpp/h}` files. Additionally, instead of copying `ReadKeyValue`, I've exposed it and `CWalletScanState` in `walletdb.h` so that they can be called from `salvage.cpp`. Lastly these changes are done in separate commits to make it easier to review.

Empact commented on src/wallet/walletdb.cpp:
```diff
@@ -924,7 +924,7 @@ bool WalletBatch::VerifyEnvironment(const fs::path& wallet_path, bilingual_str&
 
 bool WalletBatch::VerifyDatabaseFile(const fs::path& wallet_path, std::vector<bilingual_str>& warnings, bilingual_str& errorStr)
 {
-    return BerkeleyBatch::VerifyDatabaseFile(wallet_path, warnings, errorStr, WalletBatch::Recover);
```
`WalletBatch::Recover` two-argument version is unused as of commit: 1372aba35fa91dd2fb4103ed7081a057f9794564

Empact replied:
Nevermind, I see you got it in 61a1e5702d0dfd3cfb6d4deb28df516a7243e196

Empact commented on src/wallet/salvage.cpp:
```diff
@@ -0,0 +1,146 @@
+            // Required in LoadKeyMetadata():
+            LOCK(dummyWallet.cs_wallet);
+            fReadOK = ReadKeyValue(&dummyWallet, ssKey, ssValue,
+                                   dummyWss, strType, strErr);
```
Given the `CWalletScanState` isn't used, how about exposing a more limited version of `ReadKeyValue` instead?

achow101 replied:
I've done that instead.

Sjors commented on src/wallet/db.cpp:
```diff
@@ -419,18 +412,8 @@ bool BerkeleyBatch::VerifyDatabaseFile(const fs::path& file_path, std::vector<bi
-        {
-            errorStr = strprintf(_("%s corrupt, salvage failed"), walletFile);
+        if (!env->Verify(walletFile)) {
+            errorStr = strprintf(_("%s corrupt. Try using the wallet tool bitcoin-wallet to salvage"), walletFile);
```
In 1372aba35fa91dd2fb4103ed7081a057f9794564: maybe add "or restore from a backup."

achow101 replied:
Done

Sjors reviewed (commented):
Concept ACK. Needs a release note.

Travis is unhappy (as is my local build):
```
In file included from ./wallet/scriptpubkeyman.h:16:
3519./wallet/walletdb.h:314:112: error: member access into incomplete type 'CWallet'
3520             CWalletScanState &wss, std::string& strType, std::string& strErr) EXCLUSIVE_LOCKS_REQUIRED(pwallet->cs_wallet);
```

#10540 makes the case for running in aggressive mode by default and having  an option to run in aggressive mode. So maybe drop b816866e6756cb3800d51b8e943e5c408f5e6e0f?

The test problem described in https://github.com/bitcoin/bitcoin/issues/7463#issuecomment-306336149 seems to be specific to aggressive mode (cc @jnewbery). Perhaps we can at least add a test for non-aggressive mode? The original test wasn't very interesting. Is there a way to programatically damage a wallet, or can we add a wallet payload that needs salvaging to the test suite? That's useful for review as well.

jnewbery commented:
Concept ACK

Empact commented on src/wallet/wallettool.cpp:
```diff
@@ -103,6 +104,11 @@ static void WalletShowInfo(CWallet* wallet_instance)
     tfm::format(std::cout, "Address Book: %zu\n", wallet_instance->m_address_book.size());
 }
 
+static bool SalvageWallet(const fs::path path)
```
nit: could pass this through by reference

achow101 replied:
Done

Empact commented on src/wallet/db.cpp:
```diff
@@ -419,18 +412,8 @@ bool BerkeleyBatch::VerifyDatabaseFile(const fs::path& file_path, std::vector<bi
     if (fs::exists(walletDir / walletFile))
     {
         std::string backup_filename;
```
nit: this is no longer used

achow101 replied:
Removed

Empact commented on src/wallet/db.cpp:
```diff
@@ -419,18 +412,8 @@ bool BerkeleyBatch::VerifyDatabaseFile(const fs::path& file_path, std::vector<bi
-        if (r == BerkeleyEnvironment::VerifyResult::RECOVER_FAIL)
-        {
-            errorStr = strprintf(_("%s corrupt, salvage failed"), walletFile);
+        if (!env->Verify(walletFile)) {
```
nit: How about returning the `verify` result so we can log the error:
https://docs.oracle.com/cd/E17276_01/html/api_reference/C/dbverify.html

achow101 replied:
I don't think the results of `verify` are meaningful.

Empact commented on src/wallet/db.cpp:
```diff
@@ -342,8 +348,55 @@ bool BerkeleyBatch::Recover(const fs::path& file_path, void *callbackDataIn, boo
-    std::vector<BerkeleyEnvironment::KeyValPair> salvagedData;
-    bool fSuccess = env->Salvage(newFilename, salvagedData);
+    std::vector<KeyValPair> salvagedData;
+    bool fSuccess = false; //env->Salvage(newFilename, salvagedData);
```
nit: I'm not sure this comment is helpful

achow101 replied:
Removed

Empact commented:
Concept ACK

Sjors commented:
```
wallet/walletdb.cpp:594:12: error: calling function 'ReadKeyValue' requires holding mutex 'pwallet->cs_wallet' exclusively [-Werror,-Wthread-safety-analysis]
1647    return ReadKeyValue(pwallet, ssKey, ssValue, dummy_wss, strType, strErr);
```

ryanofsky reviewed (approved):
Code review ACK 5276a811e1f42eea2a6c6b679a3f1b418e414d34. Beautiful job with this PR! Not only are the changes much easier to review than last time I looked at this (my `--color-diff=dimmed_zebra` pager thanks you), but the code cleanups are fantastic: getting rid of callbacks inside of callbacks and void pointers, and weird return values, and unused options, and randomly placed static methods, and methods that shouldn't be methods, and so on.

Thank you for taking this on and being so responsive to suggestions.

ryanofsky commented:
To all reviewers who concepted ACKed here, I'd encourage reviewing the code, too. This PR is really well structured and easy to understand and after the first commit is just a series of moves and simplifications with no other changes. Very easy.

MarcoFalke commented:
> To all reviewers who concepted ACKed here

Ok, I hereby commit to reviewing this when the previous review comments have been addressed.

achow101 committed:
Add release notes about salvage changes

achow101 commented on src/wallet/db.h:
```diff
@@ -67,15 +67,6 @@ class BerkeleyEnvironment
 
     bool Verify(const std::string& strFile);
-    /**
-     * Salvage data from a file that Verify says is bad.
-     * fAggressive sets the DB_AGGRESSIVE flag (see berkeley DB->verify() method documentation).
-     * Appends binary key/value pairs to vResult, returns true if successful.
-     * NOTE: reads the entire database into memory, so cannot be used
-     * for huge databases.
```
I've added a rewording of this comment back in.

MarcoFalke commented on src/wallet/db.cpp:
```diff
@@ -342,8 +348,61 @@ bool BerkeleyBatch::Recover(const fs::path& file_path, void *callbackDataIn, boo
+     * for huge databases.
+     */
+    std::vector<KeyValPair> salvagedData;
+    bool fSuccess = false;
```
in commit 6c3afd67d2:

Why is this set all the way up here? The outcome is only needed to remember whether the end of the stream was reached or not. I think it should be defined after the last write to `keyHex`, but still before this code block:

Also, you could leave it uninitialized because every if branch will initialize it. If none does the initialization, at least a sanitizer could yell at us and hint at the bug.
```cpp
    bool fSuccess;
    if (keyHex != DATA_END) {
        LogPrintf("Salvage: WARNING: Unexpected end of file while reading salvage output.\n");
        fSuccess = false;
    } else {
        fSuccess = (result == 0);
    }

achow101 replied:
It was to reduce the amount of changes being done in a move commit.

achow101 replied:
Changed.

MarcoFalke commented on src/wallet/walletdb.h:
```diff
@@ -294,4 +294,7 @@ class WalletBatch
 void MaybeCompactWalletDB();
 
+//! Unserialize a give Key-Value pair and load it into the wallet
+bool ReadKeyValue(CWallet* pwallet, CDataStream& ssKey, CDataStream& ssValue, std::string& strType, std::string& strErr);
```
in commit 6d05dff02b:

wallet is a reference to an object (like all other params in this function), so it should probably be passed as one.

Not sure about the lock annotation, but it is a bit confusing to lock cs_wallet, and then call a function that locks it again immediately after

```suggestion
bool ReadKeyValue(CWallet& wallet, CDataStream& ssKey, CDataStream& ssValue, std::string& strType, std::string& strErr) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```

achow101 replied:
The annotation causes a compile error as `CWallet` is not yet defined.

MarcoFalke replied:
Fair enough

MarcoFalke commented on src/wallet/walletdb.cpp:
```diff
@@ -902,8 +909,7 @@ bool WalletBatch::RecoverKeysOnlyFilter(void *callbackData, CDataStream ssKey, C
     {
```
in commit 6d05dff02be9e113ac6f835:

The now unused `CWalletScanState dummyWss;` a few lines up should now be removed.

achow101 replied:
Removed

maflcko reviewed (approved):
ACK 94541050b8, only found style nits 💉

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 94541050b8, only found style nits 💉
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhGmQv+Mh8MY93W8aKewBmutaNXryt7hD5cskayCFIuYmD79NAthz5T6LRPVJJ6
m/Um8cpR1riPqmxQ+dlU2lsxVEcCs4NcGDZxv2vXUjxITyhjVDYX106KJMTEnDNe
uBvmsVjpFSE48Vi5gPgmzb2SDfmCSBCg//uWol/8gGmpMaX3ffdM5Xe9EjJfxC60
lVqT8mAReuQhFItYymqN8M919mddxAwJSnpRpbt8lIZpyoq2xVEAWvh7Jq1udCae
JGyWENYkkaDMBTPgbdfPvTJNlkvsBkyVsXicd42PXnT7N6d18w19LgXZYcClp5N4
QNiY7izurFhAfW4ujiWJzYenPNGRFK7xlqy/t+BZbfVXaNNDB3gTmqp4FGQ13rSE
0vyFZQ69QGI5e3UnSBCFWCnYyDHpItkSQ2CyXMRCsnLBMAZjUWkfOhbhZHusQxEA
DubuoJmYJt7T9xs9vqARAMEbjW6tEMQizZxl7GGXlOzzZmv3hNLdtM6V3U99U94S
28quPr/u
=Ghf5
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `fd61ba44cfca59f816c1a6d9c69adc6b6e328cc4036fc94ab1ab7e6e91ee289d  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108fd61ba44cfca59f816c1a6d9c69adc6b6e328cc4036fc94ab1ab7e6e91ee289df010510683e50e2e39056a6080e2ba7385c008fff01082cef8b81d85e161d9e2ccb04d03e09c08f1045ec7d3a3f00886042e830dddc5120083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010990b0596dc5b81233db4b30f3e32ddbc08f1045ec7d3a3f0080c79ba340871ac420083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010a71e861498dbba960616080d2b85572908f1045ec7d3a3f008b06eed478545f5fb0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f010bd31fd29e4db5365e2ff58fcc37ce27208f1045ec7d3a3f008f35c7874ddb79ea80083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6d)

</details>

ryanofsky reviewed (approved):
Code review ACK 94541050b84e166b6fcbaf6c30b263903dcc25f5. Changes since last review just rebase, and adding code comment and release note

Empact commented on src/wallet/db.cpp:
```diff
@@ -410,100 +334,23 @@ bool BerkeleyBatch::VerifyEnvironment(const fs::path& file_path, bilingual_str&
 }
 
-bool BerkeleyBatch::VerifyDatabaseFile(const fs::path& file_path, std::vector<bilingual_str>& warnings, bilingual_str& errorStr, BerkeleyEnvironment::recoverFunc_type recoverFunc)
+bool BerkeleyBatch::VerifyDatabaseFile(const fs::path& file_path, std::vector<bilingual_str>& warnings, bilingual_str& errorStr)
```
This `warnings` argument is now unused.

achow101 replied:
Removed

jonatack commented on src/wallet/salvage.cpp:
```diff
@@ -0,0 +1,150 @@
+    ptxn->commit(0);
+    pdbCopy->close(0);
+
+    return fSuccess;
```
35a1f948 `RecoverDatabaseFile` contains a fair amount of code to set and return bool `fSuccess`, but the value returned doesn't appear to be used. Is this logic needed?

achow101 replied:
The return value is being checked now.

jonatack reviewed (commented):
Concept ACK/almost ACK 94541050b84e16. Code review, builds are clean and tests are green, but attempting to run `bitcoin-wallet -regtest -wallet=test salvage` returns `DB_ENV->dbrename: method not permitted before handle's open method`. The same command with `info` runs fine. I've never used the salvage command before, so help using it is welcome. Functional tests would be really good here. A few comments below; feel free to ignore any nits.

Edit: re-attempt with a new wallet
```bash
$ bitcoin-wallet -regtest -wallet=new create
Topping up keypool...
Wallet info
===========
Encrypted: no
HD (hd seed available): yes
Keypool Size: 2000
Transactions: 0
Address Book: 0
$ bitcoin-wallet -regtest -wallet=new info
Wallet info
===========
Encrypted: no
HD (hd seed available): yes
Keypool Size: 2000
Transactions: 0
Address Book: 0
$ bitcoin-wallet -regtest -wallet=new salvage
DB_ENV->dbrename: method not permitted before handle's open method
```

achow101 committed:
Add basic test for bitcoin-wallet salvage

achow101 commented on test/functional/tool_wallet.py:
```diff
@@ -203,6 +203,10 @@ def test_getwalletinfo_on_different_wallet(self):
         self.log.debug('Wallet file shasum unchanged\n')
 
+    def test_salvage(self):
+        # TODO: We should implement a test for this. But such a test should be after we fix https://github.com/bitcoin/bitcoin/issues/7463
```
I've added a very basic test that  just checks it runs and exits without error on a new wallet.

achow101 commented on src/wallet/wallettool.cpp:
```diff
@@ -123,10 +130,15 @@ bool ExecuteWalletToolFunc(const std::string& command, const std::string& name)
+            WalletShowInfo(wallet_instance.get());
+            wallet_instance->Flush(true);
+        } else if (command == "salvage") {
+            SalvageWallet(path);
```
Changed to return it.

achow101 commented on src/wallet/db.cpp:
```diff
@@ -410,100 +334,23 @@ bool BerkeleyBatch::VerifyEnvironment(const fs::path& file_path, bilingual_str&
-        {
-            errorStr = strprintf(_("%s corrupt, salvage failed"), walletFile);
+        if (!env->Verify(walletFile)) {
+            errorStr = strprintf(_("%s corrupt. Try using the wallet tool bitcoin-wallet to salvage or restoring a backup"), walletFile);
```
Done

achow101 commented on src/wallet/salvage.cpp:
```diff
@@ -0,0 +1,150 @@
+
+    /**
+     * Salvage data from a file. The DB_AGGRESSIVE flag is being used (see berkeley DB->verify() method documentation).
+     * key/value pairs are appendedto salvagedData which are then written out to a new wallet file.
```
Done

achow101 commented on src/wallet/walletdb.h:
```diff
@@ -294,4 +288,7 @@ class WalletBatch
 //! Compacts BDB state so that wallet.dat is self-contained (if there are changes)
 void MaybeCompactWalletDB();
 
+//! Unserialize a give Key-Value pair and load it into the wallet
```
Done

achow101 commented on src/wallet/wallettool.cpp:
```diff
@@ -7,6 +7,7 @@
 #include <util/translation.h>
 #include <wallet/wallet.h>
 #include <wallet/walletutil.h>
+#include <wallet/salvage.h>
```
Done

achow101 commented:
> but attempting to run `bitcoin-wallet -regtest -wallet=test salvage` returns `DB_ENV->dbrename: method not permitted before handle's open method`.

Fixed. Turns out we still have to go through the whole `WalletDatbase::Create` and `VerifyEnvironment` song and dance before recovery can be done.

Also added a basic functional test to check that salvage on a new wallet does not have issues. It does not check whether salvaged actually worked or did anything useful.

jonatack commented:
Thanks -- at 84ae0578b6c the bitcoin-wallet salvage command now runs and exits. Yay! but it doesn't leave any feedback. A message that the wallet was successfully salvaged might be reassuring?

jonatack commented:
The new test passes at both 9454105 and 84ae057 so as a sanity check I'd agree there is room for improvement, if possible.

jonatack commented on src/wallet/wallettool.cpp:
```diff
@@ -103,6 +104,27 @@ static void WalletShowInfo(CWallet* wallet_instance)
+    try {
+        WalletBatch::VerifyEnvironment(path, error_string);
+    } catch (const fs::filesystem_error& e) {
+        error_string = Untranslated(strprintf("Error loading wallet. %s", fsbridge::get_filesystem_error_message(e)));
```
```suggestion
        error_string = Untranslated(strprintf("Error loading wallet %s\n", fsbridge::get_filesystem_error_message(e)));
```

jonatack commented on src/wallet/wallettool.cpp:
```diff
@@ -103,6 +104,27 @@ static void WalletShowInfo(CWallet* wallet_instance)
+        error_string = Untranslated(strprintf("Error loading wallet. %s", fsbridge::get_filesystem_error_message(e)));
+    }
+    if (!error_string.original.empty()) {
+        tfm::format(std::cerr, "Failed to open wallet for salvage :%s\n", error_string.original);
```
```suggestion
        tfm::format(std::cerr, "Failed to open wallet for salvage: %s\n", error_string.original);
```

achow101 replied:
Will fix if I have to push another change.

MarcoFalke commented:
re-ACK 84ae0578b6 🏉

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 84ae0578b6 🏉
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjebwv/XkWDTfbKudR9e0H8atjyYFTG7ra0mC6wBPs6Apbj4CIhGGrrVeT/zEYu
PUPAUO6/RJVD+nBPQTCajQ+/a7kkaRzR9W+Huu+1xXXJEbTzCME7Zld8uMMY750F
9X9VSKOR9JAqmaICejkZ6bk6lMY9HL2+VumYejhYKX0UTYiGAOspBX0hND68EFuR
eSh+OF5BnFj0g89kiZpp8QNDZGXAQ94Tl9tOV5SSD8I941Ncwgnt0LG20hTGOZ04
XN19zVW+f27yN/SuHO9powmwwIq/G3urif6eCnRxDHSGkc+9r/xa4SPXnv/SRcxn
i82alb+6+4N0UoNIr3gPtaxZQ/fhGHiWO9Q8s+p+/Tk2gK4ssjobBwrc4uAMhVmN
dmmQSQPmaR5Hi0PLKtiM3z98lXI29kbUxk/51vEhLH4sNs1i0TyEcwLgMcezO5/q
RIYhBStpCVQtJ+LxVUzB0zHK+A/jCM9Q56aw7ZS0ly6LHwmVPD1VvzNDQKhCWUpm
AANRPTgR
=FqzM
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `c579e4ed85c7c252470d4ad60a04790a109bf7cc4aabcc2e2094bbf1b5c53880  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108c579e4ed85c7c252470d4ad60a04790a109bf7cc4aabcc2e2094bbf1b5c53880f010098a8b7fe7d98d8b8e106f8d6e137d1708fff01003e2f488e533a9a6605729a725cd86c908f1045ecd02bcf0089fd024852dea29a70083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0101c2b7ee715384a6658739c76fa986b5008f1045ecd02bbf0085f8a136f93760fb80083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0103cb2098421d9043bf95353eb0b7888b808f1045ecd02bbf008f47a3709beb4e57b0083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f01068833b45dd652a82f027ec23c86b254708f1045ecd02bbf008a3bafb1dfada5e5f0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

practicalswift commented:
Concept ACK: thanks for doing this! :)

Empact commented:
Code Review ACK https://github.com/bitcoin/bitcoin/pull/18918/commits/84ae0578b6c68dda145ca65fef510ce0fdac0d7b

ryanofsky reviewed (approved):
Code review ACK 84ae0578b6c68dda145ca65fef510ce0fdac0d7b. Lot of small changes since previous review: added verify step before salvage, added basic test in new commit, removed unused scanstate variable and warnings parameter, tweaked various comments and strings, moved fsuccess variable declaration

meshcollider reviewed (commented):
Concept / light code review ACK 84ae0578b6c68dda145ca65fef510ce0fdac0d7b
